source/common/quic/envoy_quic_stream.cc

```diff
@@ -102,10 +102,85 @@ void EnvoyQuicStream::encodeTrailersImpl(spdy::Http2HeaderBlock&& trailers) {
   onLocalEndStream();
 }
 
-void EnvoyQuicStream::encodeMetadata(const Http::MetadataMapVector& /*metadata_map_vector*/) {
-  // Metadata Frame is not supported in QUICHE.
-  ENVOY_STREAM_LOG(debug, "METADATA is not supported in Http3.", *this);
-  stats_.metadata_not_supported_error_.inc();
+std::unique_ptr<Http::MetadataMap>
+EnvoyQuicStream::metadataMapFromHeaderList(const quic::QuicHeaderList& header_list) {
+  auto metadata_map = std::make_unique<Http::MetadataMap>();
+  for (const auto& [key, value] : header_list) {
+    (*metadata_map)[key] = value;
+  }
+  return metadata_map;
+}
+
+namespace {
+
+// Returns a new `unique_ptr<char[]>` containing the characters copied from `str`.
+std::unique_ptr<char[]> dataFromString(const std::string& str) {
+  auto data = std::make_unique<char[]>(str.length());
+  memcpy(&data[0], str.data(), str.length()); // NOLINT(safe-memcpy)
+  return data;
+}
+
+void serializeMetadata(const Http::MetadataMapPtr& metadata, quic::QuicStreamId id,
+                       absl::InlinedVector<quiche::QuicheMemSlice, 2>& slices) {
+  quic::NoopDecoderStreamErrorDelegate decoder_stream_error_delegate_;
+  quic::QpackEncoder qpack_encoder_(&decoder_stream_error_delegate_,
```

Comment:
Variables at function scope shouldn't be suffixed with underscore.

Reply:
Whoops, thanks! (cut-and-paste error from an earlier version where the encode was a member variable not a local variable).